enclave-runtime/src/rpc/author/mod.rs

```diff
@@ -229,45 +269,28 @@ where
 	}
 
 	fn watch_top(&self, ext: Vec<u8>, shard: ShardIdentifier) -> FutureResult<TxHash<P>, RpcError> {
-		// check if shard already exists
-		if !state::exists(&shard) {
-			//FIXME: Should this be an error? -> Issue error handling
-			return Box::pin(ready(Err(ClientError::InvalidShard.into())))
-		}
-		// decrypt call
-		let rsa_key = Rsa3072Seal::unseal().unwrap();
-		let request_vec = match rsa_key.decrypt(&ext.as_slice()) {
-			Ok(req) => req,
-			Err(_) => return Box::pin(ready(Err(ClientError::BadFormatDecipher.into()))),
-		};
-		// decode call
-		let stf_operation = match TrustedOperation::decode(&mut request_vec.as_slice()) {
-			Ok(op) => op,
-			Err(_) => return Box::pin(ready(Err(ClientError::BadFormat.into()))),
-		};
-		//let best_block_hash = self.client.info().best_hash;
-		// dummy block hash
-		let best_block_hash = Default::default();
-		Box::pin(
-			self.pool
-				.submit_and_watch(
-					&generic::BlockId::hash(best_block_hash),
-					TX_SOURCE,
-					stf_operation,
-					shard,
-				)
-				.map_err(|e| {
-					StateRpcError::PoolError(
-						e.into_pool_error()
-							.map(Into::into)
-							.unwrap_or_else(|_e| PoolError::Verification),
-					)
-					.into()
-				}),
-		)
+		self.process_top(ext, shard, TopSubmissionMode::SubmitWatch)
 	}
+}
+
+impl<'a, P> OnBlockCreated for Author<'a, P>
+where
+	P: TrustedOperationPool + Sync + Send + 'static,
+{
+	type Hash = P::Hash;
 
-	/*	fn unwatch_extrinsic(&self, _metadata: Option<Self::Metadata>, id: SubscriptionId) -> Result<bool> {
-		Ok(self.subscriptions.cancel(id))
-	}*/
+	fn on_block_created(&self, hashes: &[Self::Hash], block_hash: SidechainBlockHash) {
+		self.pool.on_block_created(hashes, block_hash)
+	}
+}
+
+impl<'a, P> SendState for Author<'a, P>
+where
+	P: TrustedOperationPool + Sync + Send + 'static,
+{
+	type Hash = P::Hash;
+
+	fn send_state(&self, hash: Self::Hash, state_encoded: Vec<u8>) -> Result<()> {
+		self.pool.rpc_send_state(hash, state_encoded).map_err(|e| e.into())
+	}
```

Comment:
Implement the `on_block_created` and `send_state` trait, basically just a facade for the pool.